### PR TITLE
Provide callback to pass client cert distinguished names to application

### PR DESCRIPTION
**Pull request by ebejan**
Add two libcurl APIs. One API registers a callback. The other registers a pointer to user data; this be passed to the callback. The callback and user data apply to a single http request; the user data points to data specific to this http request. This is the patterns of other callbacks in the libcurl.

The callback supports the interactive selection of client-side certificates. TLS certificate requests include a set of issuers (distinguished names); the client certificate's chain of trust should include one of these distinguished names. When a certificate request is received a handle to the request's distinguished names and the user data are passed to the callback.

When the callback returns if the certificate is now set we need to acquire a new Windows credential handle.

Preivously the code was in process of reviewing at curl#6718

### CODE REVIEW

**Comment by piru** [comment]
The biggest issue I have with this feature is the fact that is not backend agnostic. Having backend specific callbacks is cumbersome and I'd try to avoid it if only possible.

The actual callback data is a list of strings (the names of CAs that the server trust.  IMHO this data should be converted to curl_slist and that passed to the callback function instead of void *. Then each SSL backend would do whatever they need to do to convert the client request to this list when implementing he functionality.

File: lib/vtls/schannel.c [hunk 1]
```diff
@@ -414,6 +414,59 @@ get_cert_location(TCHAR *path, DWORD *store_name, TCHAR **store_path,
...
+        the application will pause the easy handle if user interaction is
+        required. */
+    if(data->set.ssl.fdistnames) {
+      result = (*data->set.ssl.fdistnames)(&dist_names,
+                                           data->set.ssl.fdistnamesp);
+      if(result != CURL_SSL_DN_FUNCTION_OK) {
+        failf(data, "dist names callback failed. status = %d, %ul ", status,
+              flags);
+         s_pSecFn->FreeContextBuffer(dist_names.aIssuers);
+        return result;
```
Comment (piru) [inline lib/vtls/schannel.c:453]: This return looks wrong. It'll return the callback return value which isn't CURLE_*

File: lib/vtls/schannel.c [hunk 2]
```diff
@@ -414,6 +414,59 @@ get_cert_location(TCHAR *path, DWORD *store_name, TCHAR **store_path,
...
+                                     &dist_names);
+  if(status != SEC_E_OK) {
+    infof(data, "QueryContextAttributes failed. status = %d, %ul ", status,
+          flags);
+    return CURLE_SSL_CONNECT_ERROR;
+  }
+  else {
+    /* Tell the application we have distinguished names. It is likely that
+        the application will pause the easy handle if user interaction is
+        required. */
```
Comment (piru) [inline lib/vtls/schannel.c:445]: How can the application pause the easy handle from this callback when the application needs to do user interaction? Does this happen without failing this specific request?

Some explanation how this is supposed to work would be in order, possibly the in callback documentation.

File: lib/vtls/schannel.c (hunk 1 above)
Comment (ebeworld) [reply lib/vtls/schannel.c:453]: Fixed it.

**Comment by ebejan** [comment]
My personal fork is messed up therefore fixing and resubmit it.

File: lib/vtls/schannel.c (hunk 2 above)
Comment (ebejan) [reply lib/vtls/schannel.c:445]: Will do.

**Pull request closed without merging**